odoo/addons/base/models/ir_qweb.py

```diff
@@ -181,10 +1604,332 @@ def _compile_directive_groups(self, el, options, indent):
+                    if {size} is not None:
+                        values[{repr(expr_as + '_last')}] = index + 1 == {size}
+                    values[{repr(expr_as + '_odd')}] = index % 2
+                    values[{repr(expr_as + '_even')}] = not values[{repr(expr_as + '_odd')}]
+                    values[{repr(expr_as + '_parity')}] = 'odd' if values[{repr(expr_as + '_odd')}] else 'even'
+            """), indent))
+
+        code.append(self._indent(f'log["last_path_node"] = {repr(options["root"].getpath(el))} ', indent + 1))
+        code.extend(content_foreach or self._indent('continue', indent + 1))
+
+        return code
+
+    def _compile_directive_as(self, el, options, indent):
+        """Compile `t-as` expressions into a python code as a list of strings.
+
+        This method only check if this attributes is on the same node of a
+         `t-foreach` attribute.
+        """
+        if not options.get('t-foreach') is None:
+            raise ValueError("t-as must be on the same node of t-foreach")
+        return []
+
+    def _compile_directive_out(self, el, options, indent):
+        """Compile `t-out` expressions into a python code as a list of
+        strings.
+
+        The code will contain evalution and rendering of the compiled value. If
+        the compiled value is None or False, the tag is not added to the render
+        (Except if the widget forces rendering or there is default content).
+        (eg: `<t t-out="my_value">Default content if falsy</t>`)
+
+        The output can have some rendering option with `t-options-widget` or
+        `t-options={'widget': ...}. At rendering time, The compiled code will
+        call ``_get_widget`` method or ``_get_field`` method for `t-field`.
+
+        A `t-field` will necessarily be linked to the value of a record field
+        (eg: `<span t-field="record.field_name"/>`), a t-out` can be applied
+        to any value (eg: `<span t-out="10" t-options-widget="'float'"/>`).
+        """
+        ttype = 't-out'
+        expr = el.attrib.pop('t-out', None)
+        if expr is None:
+            ttype = 't-field'
+            expr = el.attrib.pop('t-field', None)
+            if expr is None:
+                # deprecated use.
+                ttype = 't-esc'
+                expr = el.attrib.pop('t-esc', None)
+                if expr is None:
+                    ttype = 't-raw'
+                    expr = el.attrib.pop('t-raw')
+
+        code = self._flush_text(options, indent)
+
+        code_options = el.attrib.pop('t-consumed-options', 'None')
+        tag_open = self._compile_directive(el, options, 'tag-open', indent + 1) + self._flush_text(options, indent + 1)
+        tag_close = self._compile_directive(el, options, 'tag-close', indent + 1) + self._flush_text(options, indent + 1)
+        default_body = self._compile_directive(el, options, 'inner-content', indent + 1) + self._flush_text(options, indent + 1)
```

Comment:
Why `default_body` and not simply `body`?

Reply:
Because the body serves as the default value if `t-out` is falsy.